### ISSUE TITLE
Roberta-base-11 output differs when compiled static vs dynamic

### ISSUE THREAD

**Issue by cjvolzka** [onnx/onnx-mlir#2345]
# Description:

Compiling [roberta-base-11.onnx](https://github.com/onnx/models/blob/main/text/machine_comprehension/roberta/model/roberta-base-11.onnx) dynamically vs statically gives different output on CPU when given the same input tensor.

# Reproduce
Dynamic:
```
--O3 --EmitLib --mtriple=s390x-ibm-loz --mcpu=z16 ./roberta-base-11.onnx --onnx-op-stats TXT
Operations encountered:
-----------------------
  func.func            , 1
  func.return          , 1
  onnx.Add             , 173
  onnx.Cast            , 2
  onnx.Concat          , 50
  onnx.Constant        , 283
  onnx.ConstantOfShape , 2
  onnx.CumSum          , 1
  onnx.Dim             , 98
  onnx.Div             , 49
  onnx.Equal           , 1
  onnx.Erf             , 12
  onnx.Gather          , 4
  onnx.Gemm            , 1
  onnx.MatMul          , 96
  onnx.Mul             , 76
  onnx.Not             , 1
  onnx.ReduceMeanV13   , 50
  onnx.Reshape         , 48
  onnx.Softmax         , 12
  onnx.Sqrt            , 25
  onnx.Sub             , 26
  onnx.Tanh            , 1
  onnx.Transpose       , 48
  onnx.Unsqueeze       , 2
```

Statically:
```
--O3 --EmitLib --mtriple=s390x-ibm-loz --mcpu=z16 ./roberta-base-11.onnx --onnx-op-stats TXT --shapeInformation 0:1x1
Operations encountered:
-----------------------
  func.func          , 1
  func.return        , 1
  onnx.Add           , 160
  onnx.Cast          , 2
  onnx.Concat        , 37
  onnx.Constant      , 281
  onnx.CumSum        , 1
  onnx.Dim           , 72
  onnx.Div           , 49
  onnx.Equal         , 1
  onnx.Erf           , 12
  onnx.Gather        , 3
  onnx.Gemm          , 1
  onnx.MatMul        , 96
  onnx.Mul           , 75
  onnx.Not           , 1
  onnx.ReduceMeanV13 , 50
  onnx.Reshape       , 48
  onnx.Softmax       , 12
  onnx.Sqrt          , 25
  onnx.Sub           , 25
  onnx.Tanh          , 1
  onnx.Transpose     , 48
```

Input:
input tensor with shape (1, 1) and a value [-9]

# Expected Result
- Output tensors should be the same from either model since the same input tensor is used.
- Models pass with very low a_tol and r_tol since model was compiled for CPU
    - dynamic model passes with a_tol of 0.000002 and r_tol of 0

# Actual Results
- Static model results are notably different than dynamic model results.
    - CPU dynamic model passes with a_tol of 0.000002 and r_tol of 0 
    - CPU static model requires higher a_tol = .01 with an r_tol of 176.207779

# Notes:

- The "Operations encountered" is changing between the two
    - This might be expected since the graph optimization has more concrete information when using static shapes but I figure it worth pointing out.
    - If using `--maccel=NNPA`, the "Operations encountered" changes are more significant as more Ops go to zhigh when compiled statically
    - Compiling for NNPA also saw an even larger swing in a_tol and r_tol.
        - NNPA dynamic passes with a_tol of 0.01 and r_tol 0.006467.
        - NNPA static requires much higher a_tol of 0.01 and r_tol of 349.050293

**Comment by hamptonm1** [comment]
Okay I spoke with @chentong319 about why the added operations for `shapeInference`.... essentially they should be the same. `shapeInference` pass literally just passes the dimensions and nothing else. So for the output to look different is concerning but Tong gave a great explanation as to why this is the current behavior. I will let him explain, since he did such a great job at educating me :)

**Comment by chentong319** [comment]
The difference in the count of operations between static and dynamic shape may come from at least two sources:
1. Constant folding. If the ONNXShapeOp is called on a static shaped tensor, the output will be folded into a constant, and then may trigger further constant folding for operations that use the result ONNXShapeOp.
2. Shape inference. Some shape inference info may be expressed with ONNX ops if the shape is unknown, for better analysis and transformation.

In summary, the operation counts may be different for static and dynamic shape.
However, the outputs of the two binaries are expected to be close if they are given the same input. 
Are the input only different from input [-9] (should it be [[-9]]?)?
How about on other machines?

**Comment by hamptonm1** [comment]
@tungld did you add the code for `shapeInference` pass which essentially includes different operations compared to a dynamic compilation?

**Comment by chentong319** [comment]
> @tungld did you add the code for `shapeInference` pass which essentially includes different operations compared to a dynamic compilation?

More precisely, it should be dim analysis instead of shape inference, I guess.

**Comment by tungld** [comment]
> More precisely, it should be dim analysis instead of shape inference, I guess.

yes, It's me.

**Comment by cjvolzka** [comment]
> However, the outputs of the two binaries are expected to be close if they are given the same input.

They are the same input for both runs. At first I thought it might be an NNPA thing, but I also got notable differences in values when running on CPU dynamic vs static. Dynamically for CPU, an a_tol of 0.000002 and r_tol of 0 will pass. However for static, I have to use a_tol = .01 with an r_tol of 176.207779 to pass on the same input which is a big difference.

> Are the output only different for input [-9] (should it be [[-9]]?)?

I'm using our C++ so there's not `[[]]` issues like in python clients.

> How about on other machines?

My local machine is an M1 Apple so I don't have an easy way to try on other architectures. I'm trying to use `podman run --platform linux/amd64` on my local machine, but it's taking forever. In the past, when running on CPU the difference between s390x and x86 results have been miniscule.

**Comment by hamptonm1** [comment]
> > However, the outputs of the two binaries are expected to be close if they are given the same input.
> 
> They are the same input for both runs. At first I thought it might be an NNPA thing, but I also got notable differences in values when running on CPU dynamic vs static. Dynamically for CPU, an a_tol of 0.000002 and r_tol of 0 will pass. However for static, I have to use a_tol = .01 with an r_tol of 176.207779 to pass on the same input which is a big difference.
> 
> > Are the output only different for input [-9] (should it be [[-9]]?)?
> 
> I'm using our C++ so there's not `[[]]` issues like in python clients.
> 
> > How about on other machines?
> 
> My local machine is an M1 Apple so I don't have an easy way to try on other architectures. I'm trying to use `podman run --platform linux/amd64` on my local machine, but it's taking forever. In the past, when running on CPU the difference between s390x and x86 results have been miniscule.

If you need to test via a container... let me know. I am recreating a new container to test as of now :)

**Comment by hamptonm1** [comment]
> > More precisely, it should be dim analysis instead of shape inference, I guess.
> 
> yes, It's me.

@tungld Do you foresee us making advancements or is it necessary to keep things as is?

**Comment by hamptonm1** [comment]
Myself and @tungld talked about using other dimensions for the Roberta model to see if this behavior only occurs for dimension 1x1 (that is unfortunately not working at the moment) or if this happens for other input dimensions.

**Comment by hamptonm1** [comment]
Below is using shape-information 2x2

```
meganhampton@megans-mbp-2 build % ./Debug/bin/onnx-mlir --O3 --EmitLib --maccel=NNPA roberta-base-11.onnx --onnx-op-stats TXT --shapeInformation 0:2x2 -v
The ONNX model has 124645716 elements in its initializers. This value would be close to and greater than the number of parameters in the model. Because there is no way to exactly count the number of parameters, this value can be used to have a rough idea of the number of parameters in the model.
Operations encountered:
-----------------------
   func.func          , 1
   func.return        , 1
   onnx.Add           , 1
   onnx.Cast          , 2
   onnx.Constant      , 214
   onnx.CumSum        , 1
   onnx.Div           , 25
   onnx.Equal         , 1
   onnx.Erf           , 12
   onnx.Gather        , 3
   onnx.Mul           , 1
   onnx.NoValue       , 1
   onnx.Not           , 1
   onnx.ReduceMeanV13 , 50
   onnx.Reshape       , 132
   onnx.Sqrt          , 25
   onnx.Sub           , 25
   onnx.Transpose     , 48
  zhigh.Add           , 159
  zhigh.Div           , 24
  zhigh.MatMul        , 97
  zhigh.Mul           , 74
  zhigh.Softmax       , 12
  zhigh.Stick         , 375
  zhigh.Tanh          , 1
  zhigh.Unstick       , 173
[/Users/meganhampton/zDLC/onnx-mlir/build/]/Users/meganhampton/zDLC/llvm-project/build/bin/opt: opt -O3 --mtriple=arm64-apple-darwin22.5.0 --code-model small -o roberta-base-11.bc roberta-base-11.unoptimized.bc
```

Below is without `shape-information`:

```
meganhampton@megans-mbp-2 build % ./Debug/bin/onnx-mlir --O3 --EmitLib --maccel=NNPA roberta-base-11.onnx --onnx-op-stats TXT -v                         
The ONNX model has 124645716 elements in its initializers. This value would be close to and greater than the number of parameters in the model. Because there is no way to exactly count the number of parameters, this value can be used to have a rough idea of the number of parameters in the model.
Operations encountered:
-----------------------
   func.func            , 1
   func.return          , 1
   onnx.Add             , 99
   onnx.Cast            , 2
   onnx.Concat          , 146
   onnx.Constant        , 215
   onnx.ConstantOfShape , 2
   onnx.CumSum          , 1
   onnx.Dim             , 218
   onnx.Div             , 49
   onnx.Equal           , 1
   onnx.Erf             , 12
   onnx.Gather          , 4
   onnx.Max             , 12
   onnx.Mul             , 51
   onnx.NoValue         , 1
   onnx.Not             , 1
   onnx.ReduceMeanV13   , 50
   onnx.Reshape         , 144
   onnx.Sqrt            , 25
   onnx.Sub             , 26
   onnx.Transpose       , 48
   onnx.Unsqueeze       , 2
  zhigh.Add             , 26
  zhigh.MatMul          , 97
  zhigh.Mul             , 25
  zhigh.Softmax         , 12
  zhigh.Stick           , 283
  zhigh.Tanh            , 1
  zhigh.Unstick         , 135
[/Users/meganhampton/zDLC/onnx-mlir/build/]/Users/meganhampton/zDLC/llvm-project/build/bin/opt: opt -O3 --mtriple=arm64-apple-darwin22.5.0 --code-model small -o roberta-base-11.bc roberta-base-11.unoptimized.bc
```

**Comment by hamptonm1** [comment]
@tungld @cjvolzka  Seems like extra operations and the encounters do vary as well for other dimensions. I can test with other models too to see if this behavior occurs there as well.

**Comment by hamptonm1** [comment]
Okay when commenting out the `ConstantProp.cpp` a.k.a constant folding...it seems that for static and dynamic the amount of operations is the same when compiling as well as the values for CPU which is the expected result above. Thanks @tungld for meeting with me... will run this as well in the morning:

```
meganhampton@Megans-MacBook-Pro-2 build % ./Debug/bin/onnx-mlir --O3 --EmitLib roberta-base-11.onnx --onnx-op-stats TXT --shapeInformation 0:1x1 -v 
The ONNX model has 124645716 elements in its initializers. This value would be close to and greater than the number of parameters in the model. Because there is no way to exactly count the number of parameters, this value can be used to have a rough idea of the number of parameters in the model.
Operations encountered:
-----------------------
  func.func            , 1
  func.return          , 1
  onnx.Add             , 173
  onnx.Cast            , 2
  onnx.Concat          , 50
  onnx.Constant        , 287
  onnx.ConstantOfShape , 2
  onnx.CumSum          , 1
  onnx.Dim             , 96
  onnx.Div             , 49
  onnx.Equal           , 1
  onnx.Erf             , 12
  onnx.Gather          , 6
  onnx.Gemm            , 1
  onnx.MatMul          , 96
  onnx.Mul             , 76
  onnx.Not             , 1
  onnx.ReduceMeanV13   , 50
  onnx.Reshape         , 48
  onnx.Softmax         , 12
  onnx.Sqrt            , 25
  onnx.Sub             , 26
  onnx.Tanh            , 1
  onnx.Transpose       , 48
  onnx.Unsqueeze       , 4
[/Users/meganhampton/zDLC/onnx-mlir/build/]/Users/meganhampton/zDLC/llvm-project/build/bin/opt: opt -O3 --mtriple=arm64-apple-darwin22.5.0 --code-model small -o roberta-base-11.bc roberta-base-11.unoptimized.bc
[/Users/meganhampton/zDLC/onnx-mlir/build/]/Users/meganhampton/zDLC/llvm-project/build/bin/llc: llc -O3 --mtriple=arm64-apple-darwin22.5.0 --code-model small -filetype=obj -relocation-model=pic -o roberta-base-11.o roberta-base-11.bc
[/Users/meganhampton/zDLC/onnx-mlir/build/]/usr/bin/c++: c++ roberta-base-11.o -o roberta-base-11.so -shared -fPIC -L/Users/meganhampton/zDLC/onnx-mlir/build/Debug/lib -lcruntime
Shared library roberta-base-11.so has been compiled.
meganhampton@Megans-MacBook-Pro-2 build % ./Debug/bin/onnx-mlir --O3 --EmitLib roberta-base-11.onnx --onnx-op-stats TXT -v                         
The ONNX model has 124645716 elements in its initializers. This value would be close to and greater than the number of parameters in the model. Because there is no way to exactly count the number of parameters, this value can be used to have a rough idea of the number of parameters in the model.
Operations encountered:
-----------------------
  func.func            , 1
  func.return          , 1
  onnx.Add             , 173
  onnx.Cast            , 2
  onnx.Concat          , 50
  onnx.Constant        , 283
  onnx.ConstantOfShape , 2
  onnx.CumSum          , 1
  onnx.Dim             , 98
  onnx.Div             , 49
  onnx.Equal           , 1
  onnx.Erf             , 12
  onnx.Gather          , 4
  onnx.Gemm            , 1
  onnx.MatMul          , 96
  onnx.Mul             , 76
  onnx.Not             , 1
  onnx.ReduceMeanV13   , 50
  onnx.Reshape         , 48
  onnx.Softmax         , 12
  onnx.Sqrt            , 25
  onnx.Sub             , 26
  onnx.Tanh            , 1
  onnx.Transpose       , 48
  onnx.Unsqueeze       , 2
```

**Comment by hamptonm1** [comment]
> The difference in the count of operations between static and dynamic shape may come from at least two sources:
> 
> 1. Constant folding. If the ONNXShapeOp is called on a static shaped tensor, the output will be folded into a constant, and then may trigger further constant folding for operations that use the result ONNXShapeOp.
> 2. Shape inference. Some shape inference info may be expressed with ONNX ops if the shape is unknown, for better analysis and transformation.
> 
> In summary, the operation counts may be different for static and dynamic shape. However, the outputs of the two binaries are expected to be close if they are given the same input. Are the output only different for input [-9] (should it be [[-9]]?)? How about on other machines?

FYI- @cjvolzka So after doing further investigation... @chentong319 was correct that constant folding is the reason for the count of operations/values being off when compiling statically vs. dynamically. Above I commented out the code for constant folding and compiled the Roberta Model... we see the values are similar and the same amount of operations. I took things a step further and realized @sorenlassen just implemented a flag `--onnx-const-prop-disable-pattern` which essentially disables any pattern specified for constant folding. I used netron to narrow down which pattern was causing the issue (instead of disabling all constant folding). It seems like the culprit is Unsqueeze and here are my results below:

Statically

```
meganhampton@Megans-MacBook-Pro-2 build % ./Debug/bin/onnx-mlir --O3 --EmitLib roberta-base-11.onnx --onnx-const-prop-disable-pattern UnsqueezeofConst --onnx-const-prop-disable-pattern SplitOfConst --onnx-op-stats TXT --shapeInformation 0:1x1 -v
The ONNX model has 124645716 elements in its initializers. This value would be close to and greater than the number of parameters in the model. Because there is no way to exactly count the number of parameters, this value can be used to have a rough idea of the number of parameters in the model.
Operations encountered:
-----------------------
  func.func            , 1
  func.return          , 1
  onnx.Add             , 173
  onnx.Cast            , 2
  onnx.Concat          , 50
  onnx.Constant        , 285
  onnx.ConstantOfShape , 2
  onnx.CumSum          , 1
  onnx.Dim             , 96
  onnx.Div             , 49
  onnx.Equal           , 1
  onnx.Erf             , 12
  onnx.Gather          , 4
  onnx.Gemm            , 1
  onnx.MatMul          , 96
  onnx.Mul             , 76
  onnx.Not             , 1
  onnx.ReduceMeanV13   , 50
  onnx.Reshape         , 48
  onnx.Softmax         , 12
  onnx.Sqrt            , 25
  onnx.Sub             , 26
  onnx.Tanh            , 1
  onnx.Transpose       , 48
  onnx.Unsqueeze       , 3
[/Users/meganhampton/zDLC/onnx-mlir/build/]/Users/meganhampton/zDLC/llvm-project/build/bin/opt: opt -O3 --mtriple=arm64-apple-darwin22.5.0 --code-model small -o roberta-base-11.bc roberta-base-11.unoptimized.bc

```

Dynamically

```
meganhampton@Megans-MacBook-Pro-2 build % ./Debug/bin/onnx-mlir --O3 --EmitLib roberta-base-11.onnx --onnx-const-prop-disable-pattern UnsqueezeofConst --onnx-const-prop-disable-pattern SplitOfConst --onnx-op-stats TXT -v                         
The ONNX model has 124645716 elements in its initializers. This value would be close to and greater than the number of parameters in the model. Because there is no way to exactly count the number of parameters, this value can be used to have a rough idea of the number of parameters in the model.
Operations encountered:
-----------------------
  func.func            , 1
  func.return          , 1
  onnx.Add             , 173
  onnx.Cast            , 2
  onnx.Concat          , 50
  onnx.Constant        , 283
  onnx.ConstantOfShape , 2
  onnx.CumSum          , 1
  onnx.Dim             , 98
  onnx.Div             , 49
  onnx.Equal           , 1
  onnx.Erf             , 12
  onnx.Gather          , 4
  onnx.Gemm            , 1
  onnx.MatMul          , 96
  onnx.Mul             , 76
  onnx.Not             , 1
  onnx.ReduceMeanV13   , 50
  onnx.Reshape         , 48
  onnx.Softmax         , 12
  onnx.Sqrt            , 25
  onnx.Sub             , 26
  onnx.Tanh            , 1
  onnx.Transpose       , 48
  onnx.Unsqueeze       , 2
[/Users/meganhampton/zDLC/onnx-mlir/build/]/Users/meganhampton/zDLC/llvm-project/build/bin/opt: opt -O3 --mtriple=arm64-apple-darwin22.5.0 --code-model small -o roberta-base-11.bc roberta-base-11.unoptimized.bc
```

**Comment by hamptonm1** [comment]
I spoke with @AlexandreEichenberger about potentially tying Soren's flag. So we enable constant folding at level `03` and disable it for `O0`. I am working on that now but if anyone has thoughts or questions feel free to share :)

**Comment by hamptonm1** [comment]
@cjvolzka I am closing this issue out as resolved.... this PR fixes the issue: https://github.com/onnx/onnx-mlir/pull/2498. Use Level O0-O2 to disable constant propagation i.e. have a similar result when compiling dynamically vs. statically. Use Level O3 if you want to enable constant propagation. Thanks!
